enterprise/server/remote_execution/runner/runner.go

```diff
@@ -374,7 +375,37 @@ func (r *commandRunner) Run(ctx context.Context) *interfaces.CommandResult {
 		return r.sendPersistentWorkRequest(ctx, command)
 	}
 
-	return r.Container.Exec(ctx, command, &commandutil.Stdio{})
+	execResult := r.Container.Exec(ctx, command, &commandutil.Stdio{})
+
+	// If the runner has exceeded 90% of memory or disk usage, don't try to recycle
+	// it, because that may cause failures if it's reused
+	maxedOutStr := ""
+	for _, fsUsage := range execResult.UsageStats.GetPeakFileSystemUsage() {
+		if float64(fsUsage.UsedBytes)/float64(fsUsage.TotalBytes) >= .9 {
+			maxedOutStr += fmt.Sprintf(" %d/%d B disk used for %s", fsUsage.UsedBytes, fsUsage.TotalBytes, fsUsage.GetSource())
+		}
+	}
+	// Only firecracker containers set max memory limits on the VMs
+	if fc, ok := r.Container.Delegate.(*firecracker.FirecrackerContainer); ok {
+		usedMemoryBytes := execResult.UsageStats.GetMemoryBytes()
+		totalMemoryBytes := fc.VMConfig().GetMemSizeMb() * 1e6
+		if usedMemoryBytes >= int64(float64(totalMemoryBytes)*.9) {
+			maxedOutStr += fmt.Sprintf("%d/%d B memory used", usedMemoryBytes, totalMemoryBytes)
+		}
+	}
+
+	if maxedOutStr != "" {
+		r.doNotReuse = true
+
+		errStr := fmt.Sprintf("%v runner exceeded 90%% of memory or disk usage, not recycling: %s", r.GetIsolationType(), maxedOutStr)
+		if fc, ok := r.Container.Delegate.(*firecracker.FirecrackerContainer); ok {
+			errStr += fmt.Sprintf("\nSnapshot debug key: %s", fc.SnapshotDebugString(ctx))
+		}
+		alert.UnexpectedEvent("runner_maxed_out", errStr)
```

Comment:
@tylerwilliams just double-checking that you're OK with this since you're on call this week :)